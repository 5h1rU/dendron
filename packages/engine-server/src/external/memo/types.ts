import { URI } from "vscode-uri";

export type WorkspaceCache = {
  imageUris: URI[];
  markdownUris: URI[];
  otherUris: URI[];
  allUris: URI[];
  danglingRefs: string[];
  danglingRefsByFsPath: { [key: string]: string[] };
};

export type RefT = {
  label: string;
  ref: string;
};
