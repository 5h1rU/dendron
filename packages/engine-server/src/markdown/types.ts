import { DNoteRefLink, NotePropsV2 } from "@dendronhq/common-all";
import MDAST from "mdast";

export type DendronASTNode = MDAST.Parent & {
  notes?: NotePropsV2[];
  children: MDAST.Parent["children"] | DendronASTNode[];
};

export enum DendronASTTypes {
  WIKI_LINK = "wikiLink",
  REF_LINK = "refLink",
  PARAGRAPH = "paragraph",
}

export enum DendronASTDest {
  // MD_ENHANCED_PREVIEW = "MD_ENHANCED_PREVIEW",
  MD_REGULAR = "MD_REGULAR",
  HTML = "HTML",
}

export type DendronASTData = {
  dest: DendronASTDest;
};

export type WikiLinkNoteV4 = DendronASTNode & {
  type: DendronASTTypes.WIKI_LINK;
  value: string;
  data: WikiLinkDataV4;
};

export type WikiLinkDataV4 = {
  alias: string;
  anchorHeader?: string;
  prefix?: string;
};

export type NoteRefNoteV4 = DendronASTNode & {
  type: DendronASTTypes.REF_LINK;
  value: string;
  data: NoteRefDataV4;
};

export type NoteRefDataV4 = {
  link: DNoteRefLink;
};
