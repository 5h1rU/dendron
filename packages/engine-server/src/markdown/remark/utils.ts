import {
  CONSTANTS,
  DendronError,
  getSlugger,
  NoteChangeEntry,
  NoteProps,
  NoteUtils,
} from "@dendronhq/common-all";
import { normalizev2 } from "@dendronhq/engine-server";
import _ from "lodash";
import { Heading, Root } from "mdast";
import { Processor } from "unified";
import { Node } from "unist";
import { selectAll } from "unist-util-select";
import { VFile } from "vfile";
import { WikiLinkProps } from "../../topics/markdown";
import {
  DendronASTRoot,
  DendronASTTypes,
  NoteRefNoteV4_LEGACY,
} from "../types";

export const ALIAS_DIVIDER = "|";

export function addError(proc: Processor, err: DendronError) {
  const errors = proc.data("errors") as DendronError[];
  errors.push(err);
  proc().data("errors", errors);
}

export function getNoteOrError(
  notes: NoteProps[],
  hint: any
): { error: DendronError | undefined; note: undefined | NoteProps } {
  let error: DendronError | undefined;
  let note: NoteProps | undefined;
  if (_.isUndefined(notes)) {
    error = new DendronError({ msg: `no note found. ${hint}` });
    return { error, note };
  }
  if (notes.length > 1) {
    error = new DendronError({ msg: `multiple notes found for link: ${hint}` });
    return { error, note };
  }
  if (notes.length < 1) {
    error = new DendronError({
      msg: `no notes found for link: ${JSON.stringify(hint)}`,
    });
    return { error, note };
  }
  note = notes[0];
  return { error, note };
}

export class LinkUtils {
  static isAlias(link: string) {
    return link.indexOf("|") !== -1;
  }

  static hasFilter(link: string) {
    return link.indexOf(">") !== -1;
  }

  static parseAliasLink(link: string) {
    const [alias, value] = link.split("|").map(_.trim);
    return { alias, value: NoteUtils.normalizeFname(value) };
  }

  static parseDendronURI(linkString: string) {
    if (linkString.startsWith(CONSTANTS.DENDRON_DELIMETER)) {
      const [vaultName, link] = linkString
        .split(CONSTANTS.DENDRON_DELIMETER)[1]
        .split("/");
      return {
        vaultName,
        link,
      };
    }
    return {
      link: linkString,
    };
  }

  static parseLinkV2(linkString: string) {
    const LINK_NAME = "[^#\\|>]+";
    const re = new RegExp(
      "" +
        // alias?
        `(` +
        `(?<alias>${LINK_NAME}(?=\\|))` +
        "\\|" +
        ")?" +
        // name
        `(?<value>${LINK_NAME})` +
        // anchor?
        `(#(?<anchor>${LINK_NAME}))?` +
        // filters?
        `(>(?<filtersRaw>.*))?`,
      "i"
    );
    const out = linkString.match(re);
    if (out) {
      let { alias, value, anchor } = out.groups as any;
      let vaultName: string | undefined;
      ({ vaultName, link: value } = this.parseDendronURI(value));
      if (!alias) {
        alias = value;
      }
      alias = _.trim(alias);
      value = _.trim(value);
      return { alias, value, anchorHeader: anchor, vaultName };
    } else {
      return null;
    }
  }

  static parseLink(linkMatch: string) {
    linkMatch = NoteUtils.normalizeFname(linkMatch);
    let out: WikiLinkProps = {
      value: linkMatch,
      alias: linkMatch,
    };
    if (LinkUtils.isAlias(linkMatch)) {
      out = LinkUtils.parseAliasLink(linkMatch);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#").map(_.trim);
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!LinkUtils.isAlias(linkMatch)) {
        out.alias = value;
      }
    }
    return out;
  }
}

export class RemarkUtils {
  static oldNoteRef2NewNoteRef(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as DendronASTRoot;
        //@ts-ignore
        let notesRefLegacy: NoteRefNoteV4_LEGACY[] = selectAll("refLink", root);
        const slugger = getSlugger();
        notesRefLegacy.map((noteRefLegacy) => {
          // @ts-ignore;
          noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
          const { anchorStart, anchorEnd } = noteRefLegacy.data.link.data;
          if (anchorStart) {
            noteRefLegacy.data.link.data.anchorStart = normalizev2(
              anchorStart,
              slugger
            );
          }
          if (anchorEnd) {
            noteRefLegacy.data.link.data.anchorEnd = normalizev2(
              anchorEnd,
              slugger
            );
          }
        });
        changes.push({
          note,
          status: "update",
        });
        // if (out2.length > 0) {
        //   // const noteRefLegacy = root.children.splice(idx, 1)[0] as unknown as  NoteRefNoteV4_LEGACY;
        //   const noteRefLegacy = out2[0];
        //   // root.children[idx] as unknown as NoteRefNoteV4_LEGACY;
        //   // const noteRefNew: NoteRefNoteV4= {
        //   //   type: DendronASTTypes.REF_LINK_V2,
        //   //   data: {
        //   //     link: noteRefLegacy.data.link,
        //   //   },
        //   //   value: noteRefLegacy.value
        //   // }
        //   // @ts-ignore;
        //   noteRefLegacy.type = DendronASTTypes.REF_LINK_V2;
        //   // // @ts-ignore
        //   // root.children.splice(idx, 1, noteRefNew);
        //   changes.push({
        //     note,
        //     status: "update",
        //   });
        // }
      };
    };
  }

  static h1ToTitle(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children.splice(idx, 1)[0] as Heading;
          if (head.children.length === 1 && head.children[0].type === "text") {
            note.title = head.children[0].value;
          }
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }

  static h1ToH2(note: NoteProps, changes: NoteChangeEntry[]) {
    return function (this: Processor) {
      return (tree: Node, _vfile: VFile) => {
        let root = tree as Root;
        const idx = _.findIndex(
          root.children,
          (ent) => ent.type === "heading" && ent.depth === 1
        );
        if (idx >= 0) {
          const head = root.children[idx] as Heading;
          head.depth = 2;
          changes.push({
            note,
            status: "update",
          });
        }
      };
    };
  }
}
