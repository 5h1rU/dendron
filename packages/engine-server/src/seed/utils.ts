import { DendronError, ERROR_STATUS, SeedConfig } from "@dendronhq/common-all";
import { simpleGit } from "@dendronhq/common-server";
import fs from "fs-extra";
import _ from "lodash";
import path from "path";
import { WorkspaceService } from "../workspace";
import { DEFAULT_SEED_PUBLISHER } from "./constants";

export class SeedUtils {
  static genDefaultConfig(opts: {
    seed?: SeedConfig;
    name: string;
  }): SeedConfig {
    return _.defaults(opts.seed || {}, {
      name: opts.name,
      publisher: DEFAULT_SEED_PUBLISHER,
      description: "a seed waiting to sprout",
      license: "CC BY 4.0",
      repository: {
        type: "git" as const,
        url: "",
      },
      root: "vault",
    });
  }

  static async clone({
    wsRoot,
    config,
  }: {
    config: SeedConfig;
    wsRoot: string;
  }) {
    const id = this.getSeedId(config);
    const spath = this.seed2Path({ wsRoot, id });
    fs.ensureDirSync(path.dirname(spath));
    const git = simpleGit({ baseDir: wsRoot });
    await git.clone(config.repository.url, spath);
    return spath;
  }

  static getSeedId({ publisher, name }: { publisher: string; name: string }) {
    return `${publisher}.${name}`;
  }

  static seed2Path({ wsRoot, id }: { wsRoot: string; id: string }) {
    return path.join(wsRoot, "seeds", id);
  }

  static validateWorkspaceSeedConversion({ wsRoot }: { wsRoot: string }) {
    const ws = new WorkspaceService({ wsRoot });
    if (ws.config.vaults.length !== 1) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace must have exactly one vault",
        }),
      };
    }
    if (!_.isEmpty(ws.config.workspaces)) {
      return {
        error: DendronError.createFromStatus({
          status: ERROR_STATUS.INVALID_STATE,
          message: "workspace vaults not supported",
        }),
      };
    }
    return {
      error: undefined,
    };
  }
}
