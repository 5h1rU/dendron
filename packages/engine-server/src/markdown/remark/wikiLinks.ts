import _ from "lodash";
import { DendronError, NoteUtilsV2 } from "@dendronhq/common-all";
import { DEngineClientV2 } from "packages/engine-server/lib";
import { Eat } from "remark-parse";
import Unified, { Plugin } from "unified";
import { WikiLinkProps } from "../../topics/markdown";
import { DendronASTDest, WikiLinkDataV4, WikiLinkNoteV4 } from "../types";
import { addError, getNoteOrError, LinkUtils } from "./utils";
import { MDUtilsV4 } from "../utils";

export const LINK_REGEX = /^\[\[(.+?)\]\]/;

type PluginOpts = CompilerOpts;

type CompilerOpts = {
  convertObsidianLinks?: boolean;
  dest?: DendronASTDest;
  useId?: boolean;
  prefix?: string;
};

const plugin: Plugin = function (this: Unified.Processor, opts?: PluginOpts) {
  attachParser(this);
  if (this.Compiler != null) {
    attachCompiler(this, opts);
  }
};

function attachCompiler(proc: Unified.Processor, opts?: CompilerOpts) {
  const copts = _.defaults(opts || {}, {
    convertObsidianLinks: false,
    useId: false,
  });
  const Compiler = proc.Compiler;
  const visitors = Compiler.prototype.visitors;

  if (visitors) {
    visitors.wikiLink = function (node: WikiLinkNoteV4) {
      const data = node.data;
      let value = node.value;
      const { error, engine } = MDUtilsV4.getEngineFromProc(proc);
      if (error) {
        addError(proc, error);
        return "error with engine";
      }

      if (copts.useId) {
        const notes = NoteUtilsV2.getNotesByFname({
          fname: value,
          notes: engine.notes,
        });
        const { error, note } = getNoteOrError(notes, value);
        if (error) {
          addError(proc, error);
          return "error with link";
        } else {
          value = note!.id;
        }
      }

      switch (copts.dest) {
        case DendronASTDest.MD_REGULAR: {
          const alias = data.alias ? data.alias : value;
          return `[${alias}](${copts.prefix || ""}${value})`;
        }
        case DendronASTDest.HTML: {
          const alias = data.alias ? data.alias : value;
          return `[${alias}](${copts.prefix || ""}${value}.html)`;
        }
        default:
          return `unhandled case: ${copts.dest}`;
      }
    };
  }
}

function attachParser(proc: Unified.Processor) {
  function locator(value: string, fromIndex: number) {
    return value.indexOf("[", fromIndex);
  }

  function parseAliasLink(pageTitle: string) {
    const [alias, value] = pageTitle.split("|").map(_.trim);
    return { alias, value: NoteUtilsV2.normalizeFname(value) };
  }

  function parseLink(linkMatch: string) {
    linkMatch = NoteUtilsV2.normalizeFname(linkMatch);
    let out: WikiLinkProps = {
      value: linkMatch,
      alias: linkMatch,
    };
    if (LinkUtils.isAlias(linkMatch)) {
      out = parseAliasLink(linkMatch);
    }
    if (out.value.indexOf("#") !== -1) {
      const [value, anchorHeader] = out.value.split("#").map(_.trim);
      out.value = value;
      out.anchorHeader = anchorHeader;
      // if we didn't have an alias, links with a # anchor shouldn't have # portion be in the title
      if (!LinkUtils.isAlias(linkMatch)) {
        out.alias = value;
      }
    }
    return out;
  }

  function inlineTokenizer(eat: Eat, value: string) {
    const match = LINK_REGEX.exec(value);
    if (match) {
      const linkMatch = match[1].trim();
      const { value, alias, anchorHeader } = parseLink(linkMatch);
      return eat(match[0])({
        type: "wikiLink",
        value,
        data: {
          alias,
          anchorHeader,
        } as WikiLinkDataV4,
      });
    }
    return;
  }
  inlineTokenizer.locator = locator;

  const Parser = proc.Parser;
  const inlineTokenizers = Parser.prototype.inlineTokenizers;
  const inlineMethods = Parser.prototype.inlineMethods;
  inlineTokenizers.wikiLink = inlineTokenizer;
  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "wikiLink");
}

export { plugin as wikiLinks };
export { PluginOpts as WikiLinksOpts };
