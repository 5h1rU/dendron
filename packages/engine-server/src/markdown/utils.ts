import { DendronError, DEngineClientV2, DVault } from "@dendronhq/common-all";
import _ from "lodash";
import { Heading } from "mdast";
import { paragraph, root, text } from "mdast-builder";
// @ts-ignore
import katex from "rehype-katex";
import raw from "rehype-raw";
import rehypeStringify from "rehype-stringify";
import remark from "remark";
import frontmatterPlugin from "remark-frontmatter";
import math from "remark-math";
import remarkParse from "remark-parse";
import remark2rehype from "remark-rehype";
import { default as unified, default as Unified, Processor } from "unified";
import { Node, Parent } from "unist";
import { dendronPub, DendronPubOpts } from "./remark/dendronPub";
import { noteRefs, NoteRefsOpts } from "./remark/noteRefs";
import { wikiLinks, WikiLinksOpts } from "./remark/wikiLinks";
import { DendronASTData, DendronASTDest } from "./types";
// @ts-ignore
import highlight from "remark-highlight.js";

const toString = require("mdast-util-to-string");

type ProcOpts = {
  engine: DEngineClientV2;
};

export class MDUtilsV4 {
  static findIndex(array: Node[], fn: any) {
    for (var i = 0; i < array.length; i++) {
      if (fn(array[i], i)) {
        return i;
      }
    }
    return -1;
  }

  static genMDMsg(msg: string): Parent {
    return root(paragraph(text(msg)));
  }

  static getDendronData(proc: Processor) {
    return proc.data("dendron") as DendronASTData;
  }

  static setDendronData(proc: Processor, data: DendronASTData) {
    const _data = proc.data("dendron") as DendronASTData;
    return proc.data("dendron", { ..._data, ...data });
  }

  static getEngineFromProc(proc: Unified.Processor) {
    const engine = proc.data("engine") as DEngineClientV2;
    let error: DendronError | undefined;
    if (_.isUndefined(engine) || _.isNull(engine)) {
      error = new DendronError({ msg: "engine not defined" });
    }
    return {
      error,
      engine,
    };
  }

  static getNoteRefLvl(proc: Unified.Processor): number {
    return (proc.data("noteRefLvl") as number) || 0;
  }

  static setNoteRefLvl(proc: Unified.Processor, lvl: number) {
    return proc.data("noteRefLvl", lvl);
  }

  static isHeading(node: Node, text: string, depth?: number) {
    if (node.type !== "heading") {
      return false;
    }

    // wildcard is always true
    if (text === "*") {
      return true;
    }
    if (text) {
      var headingText = toString(node);
      return text.trim().toLowerCase() === headingText.trim().toLowerCase();
    }

    if (depth) {
      return (node as Heading).depth <= depth;
    }

    return true;
  }
  static proc(opts: ProcOpts) {
    const { engine } = opts;
    const errors: DendronError[] = [];
    let _proc = remark()
      .use(remarkParse, { gfm: true })
      .data("errors", errors)
      .data("engine", engine)
      .use(frontmatterPlugin, ["yaml"])
      .use({ settings: { listItemIndent: "1", fences: true, bullet: "-" } });
    return _proc;
  }

  static procFull(
    opts: ProcOpts & {
      dest: DendronASTDest;
      vault?: DVault;
      wikiLinksOpts?: WikiLinksOpts;
      noteRefOpts?: NoteRefsOpts;
      publishOpts?: DendronPubOpts;
      mathOpts?: {
        katex?: boolean;
      };
    }
  ) {
    const { dest, vault } = opts;
    let proc = this.proc(opts)
      .data("dendron", { dest, vault } as DendronASTData)
      .use(wikiLinks, opts.wikiLinksOpts)
      .use(noteRefs, { ...opts.noteRefOpts, wikiLinkOpts: opts.wikiLinksOpts });
    if (opts.mathOpts?.katex) {
      proc = proc.use(math);
    }
    // if (dest === DendronASTDest.HTML) {
    proc = proc.use(dendronPub, {
      ...opts.publishOpts,
      wikiLinkOpts: opts.wikiLinksOpts,
    });
    proc = proc.data("procFull", proc().freeze());
    return proc;
    //return proc;
  }

  static procRehype(opts: {
    proc?: Processor;
    mdPlugins?: Processor[];
    mathjax?: boolean;
  }) {
    const { proc, mdPlugins } = _.defaults(opts, { mdPlugins: [] });
    let _proc = proc || unified().use(remarkParse, { gfm: true });
    _proc = _proc.use(highlight);
    mdPlugins.forEach((p) => {
      _proc = _proc.use(p);
    });
    _proc = _proc.use(remark2rehype, { allowDangerousHtml: true }).use(raw);
    if (opts.mathjax) {
      _proc = _proc.use(katex);
    }
    return _proc.use(rehypeStringify);
  }
}
