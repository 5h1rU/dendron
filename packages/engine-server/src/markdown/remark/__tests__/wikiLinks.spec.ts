import { DEngineClientV2, WorkspaceOpts } from "@dendronhq/common-all";
import { createLogger } from "@dendronhq/common-server";
import { NoteTestUtilsV4, runEngineTestV4 } from "@dendronhq/common-test-utils";
import _ from "lodash";
import { DendronEngineV2 } from "../../../enginev2";
import { DendronASTData, DendronASTDest } from "../../types";
import { MDUtilsV4 } from "../../utils";
import { wikiLinks, WikiLinksOpts } from "../wikiLinks";

function proc(
  engine: DEngineClientV2,
  dendron: DendronASTData,
  opts?: WikiLinksOpts
) {
  return MDUtilsV4.proc({ engine })
    .data("dendron", dendron)
    .use(wikiLinks, opts);
}

describe("parse", () => {
  let engine: any;
  let dendronData = { dest: DendronASTDest.MD_REGULAR };

  test("init", () => {
    const resp = proc(engine, dendronData).parse(`[[foo.md]]`);
    expect(resp).toMatchSnapshot();
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("wikiLink");
  });

  test("doesn't parse inline code block", () => {
    const resp = proc(engine, dendronData).parse("`[[foo.md]]`");
    expect(resp).toMatchSnapshot("bond");
    // @ts-ignore
    expect(resp.children[0].children[0].type).toEqual("inlineCode");
  });
});

const createEngine = ({ vaults, wsRoot }: WorkspaceOpts) => {
  const logger = createLogger("testLogger", "/tmp/engine-server.txt");
  const engine = DendronEngineV2.createV3({ vaults, wsRoot, logger });
  return engine;
};

const basicSetup = async ({ wsRoot, vaults }: WorkspaceOpts) => {
  await NoteTestUtilsV4.createNote({
    wsRoot,
    fname: "foo",
    vault: vaults[0],
    props: { id: "foo-id" },
  });
};

const getTextString1 = (link = `[[FAQ | dendron.faq]]`) =>
  `See the ${link} for answers for`;
describe("compilev2", () => {
  const linkWithNoExtension = "[[foo]]";
  const linkWithExtension = "[[foo.md]]";
  const linkWithAlias = `[[bar|foo]]`;
  const textString1 = getTextString1();

  const expected: { [key in DendronASTDest]: any } = {
    [DendronASTDest.HTML]: {
      regular: {
        link: `[foo](foo.html)`,
      },
      regularWithExtension: {
        link: `[foo](foo.html)`,
      },
      alias: {
        link: `[bar](foo.html)`,
      },
      withId: {
        link: `[foo](foo-id.html)`,
      },
      textString1: {
        link: getTextString1(`[FAQ](dendron.faq.html)`),
      },
    },
    [DendronASTDest.MD_REGULAR]: {
      regular: {
        link: `[foo](foo)`,
      },
      regularWithExtension: {
        link: `[foo](foo)`,
      },
      alias: {
        link: `[bar](foo)`,
      },
      withId: {
        link: `[foo](foo-id)`,
      },
      textString1: {
        link: getTextString1(`[FAQ](dendron.faq)`),
      },
    },
    // [DendronASTDest.MD_ENHANCED_PREVIEW]: {
    //   regular: {
    //     link: `[foo](foo)`,
    //   },
    // },
  };

  const testCases = [
    {
      testCase: "regular",
      linkProcess: linkWithNoExtension,
      preSetupHook: undefined,
      procOpts: {},
    },
    {
      testCase: "regularWithExtension",
      linkProcess: linkWithExtension,
      preSetupHook: undefined,
      procOpts: {},
    },
    { testCase: "alias", linkProcess: linkWithAlias },
    {
      testCase: "withId",
      linkProcess: linkWithExtension,
      preSetupHook: basicSetup,
      procOpts: { useId: true },
    },
    {
      testCase: "textString1",
      linkProcess: textString1,
    },
  ];

  Object.values(DendronASTDest).map((key) => {
    const dest = key as DendronASTDest;
    describe(dest, () => {
      test.each(testCases.map((ent) => [ent.testCase, ent]))(
        "%p",
        async (_key, obj: typeof testCases[0]) => {
          const { testCase, linkProcess, preSetupHook, procOpts } = obj;
          await runEngineTestV4(
            async ({ engine }) => {
              const resp = await proc(
                engine,
                {
                  dest,
                },
                procOpts
              ).process(linkProcess);
              expect(resp).toMatchSnapshot();
              expect(_.trim(resp.toString())).toEqual(
                expected[dest][testCase].link
              );
            },
            { expect, createEngine, preSetupHook }
          );
        }
      );
    });
  });
});
