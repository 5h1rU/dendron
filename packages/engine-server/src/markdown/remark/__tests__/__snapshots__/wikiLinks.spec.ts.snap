// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`compilev2 HTML "alias" 1`] = `
VFile {
  "contents": "[bar](foo.html)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regular" 1`] = `
VFile {
  "contents": "[foo](foo.html)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo.html#one)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo.html)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq.html) for answers for
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 HTML "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id.html)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "alias" 1`] = `
VFile {
  "contents": "[[bar|foo]]
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regular" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regularWithAnchor" 1`] = `
VFile {
  "contents": "[[foo#one]]
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "regularWithExtension" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "textString1" 1`] = `
VFile {
  "contents": "See the [[FAQ|dendron.faq]] for answers for
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_DENDRON "withId" 1`] = `
VFile {
  "contents": "[[foo]]
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "alias" 1`] = `
VFile {
  "contents": "[bar](foo.md)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regular" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo.md)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq.md) for answers for
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_ENHANCED_PREVIEW "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id.md)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "alias" 1`] = `
VFile {
  "contents": "[bar](foo)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regular" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regularWithAnchor" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "regularWithExtension" 1`] = `
VFile {
  "contents": "[foo](foo)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "textString1" 1`] = `
VFile {
  "contents": "See the [FAQ](dendron.faq) for answers for
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`compilev2 MD_REGULAR "withId" 1`] = `
VFile {
  "contents": "[foo](foo-id)
",
  "cwd": "/Users/kevinlin/projects/dendronv2/dendron/packages/engine-server",
  "data": Object {},
  "history": Array [],
  "messages": Array [],
}
`;

exports[`parse doesn't parse inline code block: bond 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "inlineCode",
          "value": "[[foo.md]]",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 13,
          "line": 1,
          "offset": 12,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 13,
      "line": 1,
      "offset": 12,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;

exports[`parse init 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "data": Object {
            "alias": "foo",
            "anchorHeader": undefined,
          },
          "position": Position {
            "end": Object {
              "column": 11,
              "line": 1,
              "offset": 10,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "wikiLink",
          "value": "foo",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 11,
          "line": 1,
          "offset": 10,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 11,
      "line": 1,
      "offset": 10,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
